### PR TITLE
Added output-commodity-style option

### PR DESCRIPTION
**Pull request by alang76**
First attempt at adding an extra command line option to enable overriding output commodity styles. This should also cover setting precision as requested here: https://github.com/simonmichael/hledger/issues/1162

### CODE REVIEW

**Comment by Xitian9** [comment]
I don't think there's a need to create a new `OutputCommodityStyleOpts` datatype which is stored in `CliOpts`. This can just be put in the existing `commodity_styles_` option in `BalancingOpts` (in `InputOpts`, in `CliOpts`).

**Comment by simonmichael** [comment]
Thanks for the PR.

Does it make sense to put this in "BalancingOpts" though ? I'd rather find the most logical place to put it.

**Comment by simonmichael** [comment]
(Renaming things might be enough to make that logical place ?)

**Comment by Xitian9** [comment]
It could very easily be pulled out somewhere else. I created `BalancingOpts` just because there were two Boolean opts being passed to the balancing functions, with lots of possibility for conclusion. I bundled `commodity_styles_` with those because it was always used alongside them, but could be pulled up to `InputOpts` (or higher).

**Comment by alang76** [comment]
What is a little confusing is that `output_commodity_styles_` is essentially an output option, more akin to output-file and output-format. If we would pull it out from `InputOpts`, then we would have to pass the entire CliOpts to the balancing functions, just so it could use this other field which happens to share the same type. Could it be an alternative to wrap the balancing booleans with the needed styles map before the balancing functions are called because it's not currently specified as a command line parameter anyway?

**Comment by simonmichael** [comment]
Sorry, I haven't fully re-grokked all our options structures but @alang76's approach does seem to make sense to me. Unless there's a way to simplify further, which of course I'd be interested in.

Some UI-ish thoughts. 

`-c` is a valuable short flag but I agree with using it for this.

The flag name is long. I was thinking of consistent "--output-X" options, but we have many other options which also affect output. How about just `--commodity-style` ? Shorter and makes the `-c` more mnemonic. (But read on..)

Out of scope for this PR, but perhaps related: we would also like to be able to rewrite commodity symbols ("commodity aliases"). Simplest would be an independent option, say --commodity-alias, similar to --alias for accounts. Perhaps more convenient for users though, should it be combined it with the commodity style option ? As in, say,
```
-c/--commodity-style=[SYMBOL=](AMOUNT|SYMBOL)
```
Examples:
1. `-c 'USD 1,000.00'`      set the style for USD, as above
2. `-c '$=USD 1,000.00'`   rewrite $ to USD and set the style
3. `-c '$=USD'`   rewrite $ to USD (and use the style of USD if any, otherwise keep the style of $)
4. `-c '$='`   remove $ symbols (and use the style of "" if any, otherwise keep the style of $)
5. `-c '=$'`   add $ symbol to no-symbol amounts (and use the style of $ if any, otherwise keep the style of "")
```
Even a simpler form of this could be quite useful. If we did combine these functionalities, we might want to revisit the "--commodity-style" long flag name ? Some possibilities:

- `--commodity-style`
- `--commodity-display`
- `--commodity-as`
- `--commodity`

**Comment by simonmichael** [comment]
This is incomplete of course; I forgot about the plan of setting style for all commodities when there's no commodity symbol. We'd need to drop that or integrate with that somehow.

**Comment by simonmichael** [comment]
Testing: how come this fails ?
```
ghci> :main -f examples/sample.journal print --output-commodity-style='EUR1.000,00'
*** Exception: ParseErrorBundle {bundleErrors = TrivialError 12 (Just (Tokens ('\'' :| ""))) (fromList [Tokens (' ' :| ""),Label ('d' :| "igit"),Label ('e' :| "xponent"),Label ('l' :| "edger-style lot date"),Label ('l' :| "edger-style lot price"),Label ('t' :| "ransaction price"),EndOfInput]) :| [], bundlePosState = PosState {pstateInput = "'EUR1.000,00'", pstateOffset = 0, pstateSourcePos = SourcePos {sourceName = "", sourceLine = Pos 1, sourceColumn = Pos 1}, pstateTabWidth = Pos 8, pstateLinePrefix = ""}}
```
Error message: instead of 
```
ghci> :main -f examples/sample.journal print -c 'test'  
*** Exception: ParseErrorBundle {bundleErrors = TrivialError 6 (Just EndOfInput) (fromList [Tokens ('+' :| ""),Tokens ('-' :| ""),Label ('n' :| "umber")]) :| [], bundlePosState = PosState {pstateInput = "'test'", pstateOffset = 0, pstateSourcePos = SourcePos {sourceName = "", sourceLine = Pos 1, sourceColumn = Pos 1}, pstateTabWidth = Pos 8, pstateLinePrefix = ""}}
```
it would be good to show something like
```
--output-commodity-style could not parse "test", please use amount syntax
https://hledger.org/hledger.html#amounts
```

**Comment by simonmichael** [comment]
We'll want a test or two; either unit or functional, the latter will be easier probably.

**Comment by simonmichael** [comment]
And (as last step) we'll want to follow the [new commit conventions scheme](https://hledger.org/CONTRIBUTING.html#commit-messages), eg adding the `feat: ` prefix.

**Comment by Xitian9** [comment]
> What is a little confusing is that `output_commodity_styles_` is essentially an output option, more akin to `output_file_` and `output_format_` If we would pull it out from `InputOpts`, then we would have to pass the entire `CliOpts` to the balancing functions, just so it could use this other field which happens to share the same type. 

I agree that where the styles map lives is a bit confusing. It does seem more like an output option than an input option, but it plays two important roles in InputOpts:
1. It styles any inferred amounts (i.e. amounts left blank in the journal)
2. It determines whether amounts look zero (because we consider them balanced if they look zero; they don't need to actually be zero to arbitrary precision).

> Could it be an alternative to wrap the balancing booleans with the needed styles map before the balancing functions are called because the balancing styles map is not currently specified as a command line parameter anyway?

I don't think I fully understand how this differs from the current situation. Could you give an example?

**Comment by alang76** [comment]
Thanks @simonmichael  and @Xitian9 for your comments. I need some more time to work on the recommendations. Some upfront comments:

@simonmichael:

> UI-ish thoughts

Commodity-style looks better indeed. I like the idea of combining it the optional alias. Agree with out of scope for this PR.

> Testing: how come this fails ?

Regarding the first error, is ghci doing something funny with the quotes? The compiled binary doesn't have any issues it seems.

> Error message: instead of

I'll add a proper error message. Since you can pass the option multiple times to supply the style for multiple commodities, I guess it would be nice to list all the problematic ones

> We'll want a test or two; either unit or functional, the latter will be easier probably.

Agreed.

@Xitian9:

> It does seem more like an output option than an input option, but it plays two important roles in InputOpts:

When mentioning the output option, I was more referring to the new option for overriding the style (ie the one added in this PR). I agree that the location of the original one inside `BalancingOpts` was probably fine to be placed there. My point was that there appear  to be two different use cases: one serving as type to hold inference results in the balancing function and the other for overriding the output styles. If both use cases consumed the same option value from `CliOpts` that could be a source of confusion.

> I don't think I fully understand how this differs from the current situation. Could you give an example?

I was suggesting that the potential issue of having two use cases for the same option as mentioned above might be solved by removing the one from balancing opts altogether from the CliOpts, since it is not set from the command line at all if I am correct? If its only there to support the inference process in the balancing function, could it not be passed along to that function at a later stage, after command options have been parsed? On the other hand I guess there is no real downside to having multiple options with different purposes, which just happen to have the same type. Or maybe I'm not fully understanding :)

**Comment by simonmichael** [comment]
Oh you're probably right about ghci quoting, thanks. 

Good idea, to show the option value that failed to parse. Probably just showing the first failure is enough, that's our usual MO.

**Comment by simonmichael** [comment]
The new "check commit messages" CI step always fails on force pushes I think. A regular push of a new commit might get CI working.

The changes are looking good. We should figure out a good place to document the new option

**Comment by simonmichael** [comment]
A new section under hledger.m4.md -> OUTPUT, perhaps

**Comment by alang76** [comment]
> A regular push of a new commit might get CI working.

Ah right was wondering about that.

> A new section under hledger.m4.md -> OUTPUT, perhaps

Sounds good, let me try and propose a description.

@Xitian9, in the end I figured the semantics of the original parameter was pretty much the same as you suggested. It's pulled out at InputOpts level now.

**Comment by Xitian9** [comment]
Sounds great! At this point I think `BalancingOpts` can just be eliminated and its two remaining booleans put into `InputOpts`. If you want to do that you're welcome to, otherwise I'll submit a PR once this one has been merged.

In `CliOptions.hs` some of the fields of `CliOpts`, its default declaration, and the record update in `rawOptsToCliOpts` have gained some extra spacing that's no longer necessary.

**Comment by simonmichael** [comment]
CI should be ok with force pushes now.

Where/when to mention this in docs is a little tricky. I'm thinking

## OUTPUT

### Commodity styles

*1. new section, describing --commodity-style option as briefly as we can get away with and linking to...*

## JOURNAL FORMAT

### Amounts

#### Commodity display style

*2. At end, mention --commodity-style for overriding, link to 1*

### Declaring commodities

*3. At end, mention --commodity-style for overriding, link to 1*

I can add this if you prefer.

**Comment by alang76** [comment]
@Xitian9:  Balancing opts is gone now. Does it look OK? Was a bit unsure about the location of the defaults.

@simonmichael: Feel free to change the text to be more in line with the general style

**Comment by Xitian9** [comment]
@alang76 Sorry, by flattening `BalancingOpts` into `InputOpts`, I meant that `InputOpts` should be passed around wherever `BalancingOpts` previously were. Passing two separate booleans as arguments to the balancing functions will give us a bad case of boolean blindness, especially when the default values of each are different (e.g. what does `journalBalanceTransactions True False` do, without looking up the definitions in code?).

I would just get rid of `defaultIgnoreAssertions` and `defaultInferPrices`. It's sufficient to just put the defaults in `defaultInputOpts`.

**Comment by apauley** [comment]
Thanks for adding this @alang76, @simonmichael and @Xitian9 - this can solve an issue that an hledger-flow user have reported when converting Paypal CSV files to journals (here is a [simplified summary](https://github.com/simonmichael/hledger/issues/1375#issuecomment-876748013) for background context).

I have built this branch and tested it using `hledger print --commodity-style '€1 000.00'` on a test file - it works great like that.

Not sure if this is too big of an ask, but would it be possible to allow the commodity styles to be specified using an external file?

hledger-flow uses a file named `directives.journal` and then automatically includes it in the main journal if it exists.

This is what my directives.journal looks like:
```
commodity $1,000.00
commodity 1,000.00 USD
P 2000-01-01 USD $1

commodity €1,000.00
commodity 1,000.00 EUR
P 2000-01-01 EUR €1
```

Internally hledger-flow runs `hledger print` to convert CSV files to hledger journals, and it would be convenient to provide the same file as an option there.

Alternatively I can also extract all the commodity directives from that file, and specify them as multiple `--commodity-style` options - this can then avoid over-complicating this PR if not many people will find an external file useful.

**Comment by Xitian9** [comment]
That looks good! @simonmichael has mentioned that he'd like to reduce the use of `def` in the code base due to it interfering with IDE and inference tools, so maybe those `def :: InputOpts` could be replaced by `definputopts`.

**Comment by alang76** [comment]
@Xitian9 Yeah that's a good idea. I was a bit unfamiliar with Data.Default but according to [this](https://markkarpov.com/post/data-default.html) there are also other reasons why they are not a great idea.

@apauley Hmm that could be done I guess, but that would be really close to the hledger import behaviour wouldn't it?

**Comment by apauley** [comment]
> @apauley Hmm that could be done I guess, but that would be really close to the hledger import behaviour wouldn't it?

True, I can also switch to using `hledger import`. It behaves a bit differently, so I'll have to test that the `.latest.bank.csv` behaves well with the rest of the hledger-flow assumptions.

**Comment by alang76** [comment]
@apauley Ah yeah that's true and in fact one of the reasons I prefer print in my own workflow. Curious to know what @simonmichael and @Xitian9 think of the proposal. In the interest of gettings things done ™ maybe we could take this on in a followup PR?

**Comment by apauley** [comment]
> In the interest of gettings things done ™ maybe we could take this on in a followup PR?

Sure, a follow-up PR sounds good - assuming it should be done at all.

**Comment by simonmichael** [comment]
I squashed the current commits into one and suddenly everything became very clear - can we do that ?

Could you add the hledger-lib.cabal change in package.yaml also. (I would add it to package.yaml first and do a stack build to be certain sure all the right places in the cabal file get updated.)

I think this is looking great. We are passing more unused options in some places but the simplification seems worth it.

Is InputOptions used more in Hledger.Data or in Hledger.Read ? If the latter, how hard would it be to move it there ?

**Comment by simonmichael** [comment]
Thanks for the input @apauley. Yes let's discuss this separately (chat could be good too). A separate file normally runs into https://hledger.org/hledger.html#directives-and-multiple-files limitations, and a special case doesn't sound so good.

**Comment by apauley** [comment]
> Thanks for the input @apauley. Yes let's discuss this separately (chat could be good too).

I think I should test `hledger import`, it does what I need when it comes to the directives, and I think there are ways to ensure that the `.latest.bank.csv` plays nicely with the rest of hledger-flow. We can also chat if there are more options that you think I can consider.

**Comment by alang76** [comment]
> Is InputOptions used more in Hledger.Data or in Hledger.Read ? If the latter, how hard would it be to move it there ?

As discussed there is a situation now where stuff is used for reading, which should be in Hledger.Read, but with a dependency from the lower level Hledger.Data. Might need to take a second look at which types do what.

**Comment by simonmichael** [comment]
Thanks. Yes from chat I'm clearer that we moved InputOpts, primarily Read-related, into Data. Let's take a little more time to think on the best factoring/naming/placement here.

**Comment by Xitian9** [comment]
The only thing we need `Hledger.Data` for in the new `InputOpts` module is for
```
type StorageFormat = String
type CommodityStyles = Maybe (M.Map CommoditySymbol AmountStyle)
defcommoditystyles = Nothing
```

`defcommoditystyles` isn't pulling a lot of weight, and can probably just be replaced with `Nothing` everywhere.

I'm personally not a big fan of type synonyms (and I'm [not the only one](https://www.reddit.com/r/haskell/comments/64ui11/reasons_i_dislike_type_synonyms/)), but if we just substitute the actual types in the definition of `InputOpts` then we require no imports whatsoever and can put that module wherever we like.

**Comment by Xitian9** [comment]
Oh, I see. The problem is that `InputOpts` is now used inside `Hledger.Data.Transaction`.

Well, it may not be aesthetically pleasing with the hierarchical naming scheme, but so long as `InputOptions` is a leaf of the dependency graph, it can be imported anywhere without causing problems. There's nothing stopping us putting it in `Hledger.Read.InputOptions` and importing that in both `Hledger.Data.Transaction` and the root `Hledger.Read` module.

**Comment by alang76** [comment]
Yes that could work practically I think. I wonder though, wouldn't it be conceptually strange if Data.Transaction (which is a core module) would depend on a data structure which is part of the shell around it to expose it the UI? I keep thinking that the original `BalancingOpts` wasn't such a bad idea, except maybe that having the `InputOpts` include `BalancingOpts` mixes types defined in `Hledger.Data` and `Hledger.Read`, but that is perhaps purely an aesthetic concern.

**Comment by Xitian9** [comment]
Well, another option would be to take `transactionCheckBalanced` and `isTransactionBalanced` out of `Hledger.Data.Transactions` and `journalBalanceTransactions` and `journalCheckBalanceAssertions` out of `Hledger.Data.Journal`. They could perhaps be put somewhere like `Hledger.Read.Balancing`.

**Comment by alang76** [comment]
Hmm, maybe for balancing the conceptual separation between Read and Data isn't as large as I initially thought. I will meditate on this during my holiday :)

Thanks for the hint regarding type synonyms btw.

**Comment by Xitian9** [comment]
Enjoy your holiday, and try to find more pleasant things to meditate upon!

**Comment by Xitian9** [comment]
Hi @alang76. Is there anything I can do to help get this out the door?

**Comment by alang76** [comment]
Hi @Xitian9, going to try give this a push today.

**Comment by alang76** [comment]
Disregard above.

**Comment by alang76** [comment]
- Moved `InputOpts` to `Hledger.Read`
- Removed type synonym for the commodity styles
- Reverted to the original `BalancingOpts` (I think this is better then flattening into `InputOpts` since that requires `Hledger.Data.Transaction` to have a dependency on `InputOpts` which is now in `Hledger.Read`)
- Removed default instances for `BalancingOpts` and `InputOpts`

File: hledger/test/cli/commodity-style.test [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+2021-07-09 Euro
+    (a)    EUR 1.234,56
+
+2021-07-09 Dollar
+    (a)      $ 1,234.56
+
+$ hledger -f- print -c '1 0'
+>
+2021-07-09 no symbol
+    (a)           1 2 3
```
Comment (Xitian9) [inline hledger/test/cli/commodity-style.test:15]: I think I might need some clarification with what's going on here. Here we're changing to digit groupings of size one with a space for a ‘thousands’ separator? Odd, but reasonable.

**Comment by Xitian9** [comment]
Looks great! There are a couple of missing newlines at the end of files, but I don't think that's a blocker for merging.

File: hledger/test/cli/commodity-style.test (hunk 1 above)
Comment (Xitian9) [reply hledger/test/cli/commodity-style.test:15]: To be clear, a space for a thousands separator is the only reasonable option ^^. I meant the digit groupings of sizse one was odd.

Comment (alang76) [reply hledger/test/cli/commodity-style.test]: Yeah, although an interesting corner case, this is probably not the place to test it.

**Comment by alang76** [comment]
> Looks great! There are a couple of missing newlines at the end of files

Where are you seeing them? I'll fix it up

**Comment by Xitian9** [comment]
Github will show you with a little red circle with a line through it in the PR. Check the end of `commodity-style.test` under Files changed.

**Comment by alang76** [comment]
@Xitian9 Should be fixed now.

**Comment by simonmichael** [comment]
Looks great to me, rebased and merged as 3426030a. Thank you very much!
